README.md
# decent-unit

Decent conversion of measurement units.

[![Build Status](https://travis-ci.org/lakca/decent-unit.svg?branch=master)](https://travis-ci.org/lakca/decent-unit)
[![codecov](https://codecov.io/gh/lakca/decent-unit/branch/master/graph/badge.svg)](https://codecov.io/gh/lakca/decent-unit)

## Usage

```js
const decentUnit = require('@lakca/decent-unit')
```

## Example

```js
decentUnit(2 * 1024 ** 3) // decent result: { value: 2, unit: 'G' }

decentUnit(2 * 1024 ** 3, { fromUnit: 'G' }) // decent result: { value: 2, unit: 'E' }

decentUnit(2 * 1000 ** 3, { fromUnit: 'G', radix: 1000 }) // decent result { value: 2, unit: 'E' }

decentUnit(20000, { fromUnit: '', radix: [100, 20, 10, 2], units: ['', 'K', 'M', 'G', 'T'] }) // decent result { value: 1, unit: 'G' }
 *

decentUnit(2 * 1024 ** 3, { fromUnit: 'G', toUnit: 'P' }) // expected result { value: 2048, unit: 'P' }
```

More in [test](test/main.js).

## API

### `decentUnit(num: number, options?: object)`

#### `{T} [options.fromUnit='']`

> unit of `num`.

#### `{T} [options.toUnit]`

> target unit.

#### `{number|number[]} [options.radix=1024]`

> conversion radix.
  If is an `array`, element is radix between every nearby units,
  and filled with the last array element if length is more than 1 less than `options.units`.

#### `@param {T[]} [options.units=['', 'K', 'M', 'G', 'T', 'P', 'E']]`

> available units

See [decentUnit](index.js#L28)

## LICENSE

MIT


History.md
# 1.0.0

- custom `radix` sequence.
- custom `units` sequence.


test/main.js
const assert = require('assert')
const decentUnit = require('../index')

describe('', function() {
  describe('decent', function() {
    it('> radix', function(done) {
      assert.deepEqual(decentUnit(2 * Math.pow(1024, 3)), { value: 2, unit: 'G' }, 'default options')
      assert.deepEqual(decentUnit(2 * Math.pow(1024, 3), { fromUnit: 'K' }), { value: 2, unit: 'T' }, 'custom fromUnit')
      assert.deepEqual(decentUnit(2 * Math.pow(1000, 3), { fromUnit: 'K', radix: 1000 }), { value: 2, unit: 'T' }, 'custom radix')
      assert.deepEqual(decentUnit(2 * Math.pow(1000, 3), { fromUnit: 'K', radix: [1000, 100] }), { value: 20, unit: 'P' }, 'custom array radix')
      assert.deepEqual(decentUnit(20000, { fromUnit: '', radix: [100, 10, 2], units: ['', 'K', 'M', 'G'] }), { value: 10, unit: 'G' }, 'custom array radix')
      done()
    })
    it('> 1, < radix', function(done) {
      assert.deepEqual(decentUnit(256), { value: 256, unit: '' }, 'default options')
      assert.deepEqual(decentUnit(100, { fromUnit: 'K' }), { value: 100, unit: 'K' }, 'custom fromUnit')
      assert.deepEqual(decentUnit(100, { fromUnit: 'E', radix: 1000 }), { value: 100, unit: 'E' }, 'custom radix')
      assert.deepEqual(decentUnit(100, { fromUnit: 'E', radix: [1000, 200] }), { value: 100, unit: 'E' }, 'custom array radix')
      assert.deepEqual(decentUnit(50, { fromUnit: 'K', radix: [100, 10, 2], units: ['', 'K', 'M', 'G'] }), { value: 2.5, unit: 'G' }, 'custom array radix')
      done()
    })
    it('< 1', function(done) {
      assert.deepEqual(decentUnit(0.1), { value: 0.1, unit: '' }, 'default options')
      assert.deepEqual(decentUnit(0.1, { fromUnit: 'G' }), { value: 102.4, unit: 'M' }, 'custom fromUnit')
      assert.deepEqual(decentUnit(0.1, { fromUnit: 'T', radix: 1000 }), { value: 100, unit: 'G' }, 'custom radix')
      assert.deepEqual(decentUnit(0.1, { fromUnit: 'T', radix: [1000, 100] }), { value: 10, unit: 'G' }, 'custom array radix')
      assert.deepEqual(decentUnit(0.01, { fromUnit: 'G', radix: [100, 10, 2], units: ['', 'K', 'M', 'G'] }), { value: 20, unit: '' }, 'custom array radix')
      done()
    })
    it('out range', function(done) {
      assert.deepEqual(decentUnit(0.1), { value: 0.1, unit: '' }, 'default options')
      assert.deepEqual(decentUnit(0.1, { fromUnit: 'a', units: ['a', 'b', 'c'] }), { value: 0.1, unit: 'a' }, 'custom fromUnit')
      assert.deepEqual(decentUnit(2 * 1000, { fromUnit: 'E', radix: [1000, 100] }), { value: 2000, unit: 'E' }, 'custom radix')
      done()
    })
  })
  describe('fixed toUnit', function() {
    it('in range', function(done) {
      assert.deepEqual(decentUnit(2 * Math.pow(1024, 3), { toUnit: 'T' }), { value: 2/1024, unit: 'T' }, 'default options')
      assert.deepEqual(decentUnit(2 * Math.pow(1024, 3), { toUnit: 'T', fromUnit: 'K' }), { value: 2, unit: 'T' }, 'custom fromUnit')
      assert.deepEqual(decentUnit(0.01, { toUnit: 'K', fromUnit: 'G' }), { value: 0.01 * Math.pow(1024, 2), unit: 'K' }, 'custom fromUnit, greater than toUnit')
      assert.deepEqual(decentUnit(2 * Math.pow(1000, 3), { toUnit: 'G', fromUnit: 'K', radix: 1000 }), { value: 2000, unit: 'G' }, 'custom radix')
      assert.deepEqual(decentUnit(2 * Math.pow(1000, 3), { toUnit: 'G', fromUnit: 'K', radix: [1000, 100] }), { value: 200000, unit: 'G' }, 'custom array radix')
      done()
    })
  })
  describe('throw', function() {
    it('out range', function(done) {
      assert.throws(() => decentUnit(2, { fromUnit: 'x' }), 'default options')
      assert.throws(() => decentUnit(2, { toUnit: 'x' }), 'default options')
      done()
    })
  })
})


index.js
const UNITS = ['', 'K', 'M', 'G', 'T', 'P', 'E']
/**
 * Convert number from one to another with particular radix (and unit).
 *
 * - When no target unit provided, a decent target unit will be applied.
 *
 * - `Decent target unit`: the number after converted is greater than `1` and less than `radix`, as possible.
 *
 * @example
 *
 * decentUnit(2 * 1024 ** 3) // decent result: { value: 2, unit: 'G' }
 * decentUnit(2 * 1024 ** 3, { fromUnit: 'G' }) // decent result: { value: 2, unit: 'E' }
 * decentUnit(2 * 1000 ** 3, { fromUnit: 'G', radix: 1000 }) // decent result { value: 2, unit: 'E' }
 * decentUnit(20000, { fromUnit: '', radix: [100, 20, 10, 2], units: ['', 'K', 'M', 'G', 'T'] }) // decent result { value: 1, unit: 'G' }
 *
 * decentUnit(2 * 1000 ** 3, { fromUnit: 'G', toUnit: 'P' }) // expected result { value: 2000, unit: 'P' }
 *
 * @template {string} T
 * @param {number} num source number.
 * @param {object} [options]
 * @param {T} [options.fromUnit=''] unit of `num`.
 * @param {T} [options.toUnit] target unit.
 * @param {number|number[]} [options.radix=1024] conversion radix.
 *  - If is an `array`, element is radix between every nearby units,
 *  and filled with the last array element if length is more than 1 less than `options.units`.
 * @param {T[]} [options.units=['', 'K', 'M', 'G', 'T', 'P', 'E']]
 */
module.exports = function decentUnit(num, options) {
  options = Object.assign({ fromUnit: '', radix: 1024, units: UNITS }, options)
  const radix = Array.isArray(options.radix)
    ? options.radix.length > options.units.length - 2
      ? options.radix
      : options.radix.concat(new Array(options.units.length - 1 - options.radix.length).fill(options.radix[options.radix.length - 1]))
    : new Array(options.units.length - 1).fill(options.radix)
  const j = options.units.indexOf(options.toUnit)
  let i = options.units.indexOf(options.fromUnit)
  if (i < 0) throw new Error('Unknown fromUnit: ' + options.fromUnit)
  if (options.toUnit !== undefined && j < 0) throw new Error('Unknown toUnit: ' + options.toUnit)
  if (j < 0) {
    if (num >= radix[i]) {
      while (num >= radix[i] && i < options.units.length - 1) {
        num /= radix[i]
        i++
      }
    } else if (num < 1) {
      while (num < 1 && i > 0) {
        i--
        num *= radix[i]
      }
    }
  } else {
    if (i > j) {
      while (i > j) {
        i--
        num *= radix[i]
      }
    } else {
      while (i < j) {
        num /= radix[i]
        i++
      }
    }
  }
  return {
    value: num,
    unit: options.units[i]
  }
}
